torch/csrc/jit/pickler.h

```diff
@@ -183,19 +185,40 @@ class Pickler {
 
   const void* getPointer(const IValue& ivalue);
 
+  // Caller checks that bufferPos_ > 0
+  void flushNonEmpty() {
+    writer_(buffer_.data(), bufferPos_);
+    bufferPos_ = 0;
+  }
+
+  void flush() {
+    if (bufferPos_ != 0) {
+      flushNonEmpty();
+    }
+  }
+
   // These convert values to bytes and add them to the stack (NB: since T is to
   // the left of a '::', its type cannot be deduced by the compiler so one must
   // explicitly instantiate the template, i.e. push<int>(int) works, push(int)
   // does not)
   template <typename T>
   void push(typename std::common_type<T>::type value) {
     const char* begin = reinterpret_cast<const char*>(&value);
-    writer_(begin, sizeof(T));
+    if (bufferPos_ + sizeof(T) > buffer_.size()) {
+      flushNonEmpty();
+    }
+    memcpy(buffer_.data() + bufferPos_, begin, sizeof(T));
+    bufferPos_ += sizeof(T);
   }
 
   // Stream to write binary data to
+  // Code shouldn't call writer_ directly without first flush()ing.
   std::function<void(const char*, size_t)> writer_;
 
+  // Buffer to avoid calling a writer_ on a per-byte basis.
+  std::array<char, 256> buffer_;
```

Comment:
I think you could simplify a lot of the logic in this patch away by making this a `SmallVector<char, 256>`.  Then you just can just check `size()` rather than having to track `bufferPos_`, etc.